os/fs/smartfs/smartfs_utils.c

```diff
@@ -1127,56 +1127,26 @@ int smartfs_finddirentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s *d
 
 int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector, struct smartfs_entry_s *direntry)
 {
-	struct smart_read_write_s readwrite;
 	int ret;
 	uint16_t nextsector;
-	struct smartfs_chain_header_s *chainheader;
 
-	/* Start at the 1st sector in the parent directory */
 	/* If we are in the context of smartfs_createentry, then smartfs_find_availableentry did not find an available entry.
 	   Createentry has to chain a new nextsector to parentdirsector, and allocate the first entry.
 	 */
 
-	smartfs_setbuffer(&readwrite, parentdirsector, 0, fs->fs_llformat.availbytes, (uint8_t *)fs->fs_rwbuffer);
-	ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
-	if (ret < 0) {
-		fdbg("Unable to read parent sector\n");
-		return ret;
-	}
-
-	chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
-	nextsector = SMARTFS_NEXTSECTOR(chainheader);
-
-	if (nextsector != SMARTFS_ERASEDSTATE_16BIT) {
-		fdbg("Cannot chain new sector\n");
-		return -EIO;
-	}
 	/* Allocate a new sector and chain it to the last one */
-
 	ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
 	if (ret < 0) {
 		return ret;
 	}
 
 	nextsector = (uint16_t)ret;
 
-	/* Chain the next sector into this sector sector */
-	chainheader->nextsector[0] = (uint8_t)(nextsector & 0x00FF);
-	chainheader->nextsector[1] = (uint8_t)(nextsector >> 8);
-
-	/* Here we are writing to the current sector's chainheader which needed a new sector in the chain.
-	   Not writing to the newly chained sector, so it is completely empty as of now.
-	 */
-
-	smartfs_setbuffer(&readwrite, parentdirsector, offsetof(struct smartfs_chain_header_s, nextsector), sizeof(uint16_t), chainheader->nextsector);
-	ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
-	if (ret < 0) {
-		fdbg("Error chaining sector %d\n", nextsector);
-		return ret;
-	}
-
 	/* Automatically selecting first entry of new chained sector */
-	direntry->doffset = sizeof(struct smartfs_chain_header_s);
+	/* With createentry, offset is known to be sizeof(chainheader) automatically,
+	   Let's use direntry->doffset for saving the parent logical sector to write chainheader->nextsector later.
+	 */
+	direntry->doffset = parentdirsector;
```

Comment:
I think this is not good option.
1. Actually this only called from find_availableentry, so it already know parentdirsector.
2. There's chance that asign sector number to 'offset' will bring great confusion later. 
    This can make hard to debugging later.
3. If we use offset as a entire of given flash for file system, then it is acceptable but many things should be changed.
   for example,
```
direntry->doffset = parentdirsector * size_of_sector + target_offset
```

Reply:
How about assigning
direntry.dfirst = parentdirsector

Or you want that no attribute of direntry should be assigned?

Reply:
Resolved